### PR TITLE
V1.1

### PR DESCRIPTION
**Pull request by marijanafilipovic**
Merge 1.1 to master

### CODE REVIEW

File: src/CardRequestService.php [hunk 1]
```diff
@@ -5,50 +5,46 @@
...
 use GuzzleHttp\Exception\RequestException;
 
-class CardRequestService implements IRequestService
+class CardRequestService extends validateRequest implements IRequestService
 {
-    protected $client;
-    public function makeRequest($method, $endpointName=null, $param=null)
+    private $modelName = 'card';
+    private $client;
+    protected string $api_key;
```
Comment (davorpml) [inline src/CardRequestService.php:14]: You are using camelCase everywhere except here. Change it to `$apiKey` please.

Comment (davorpml) [inline src/CardRequestService.php:14]: You are using camelCase everywhere except here. Change it to `$apiKey` please.

File: src/CardRequestService.php [hunk 2]
```diff
@@ -5,50 +5,46 @@
...
+    protected string $method;
+    protected string $url;
+    protected mixed $param;
+    public function validateRequest($modelName, $method, $endpointName, $param=null)
```
Comment (davorpml) [inline src/CardRequestService.php:18]: Code style: add a blank line before method

File: src/CardRequestService.php [hunk 3]
```diff
@@ -5,50 +5,46 @@
...
-           // $param = 150;
-            if(strpos($endUrl, ':id') !==false && (!empty($param))){
-                $endUrl = str_replace(':id', $param,$endUrl);
-            }
-            $baseURL = env('APP_URL');
-            $url = $baseURL.$endUrl;
-            //dd($url);
-        }
-        //check response
+        $requestCheck = new validateRequest();
```
Comment (davorpml) [inline src/CardRequestService.php:20]: intendation

File: src/CardRequestService.php [hunk 4]
```diff
@@ -5,50 +5,46 @@
...
+       return $this->url = $requestCheck->validate($modelName, $method, $endpointName, $param);
```
Comment (davorpml) [inline src/CardRequestService.php:21]: add a blank line before return

File: src/CardRequestService.php [hunk 5]
```diff
@@ -5,50 +5,46 @@
...
+    }
 
+    public function makeRequest($method, $url, $param=null)
+    {
+        try{
```
Comment (davorpml) [inline src/CardRequestService.php:26]: code style: `try {`

File: src/CardRequestService.php [hunk 6]
```diff
@@ -5,50 +5,46 @@
...
                 $client = new Client();
                 $api_key = config('auth_app.AUTH-KEY');
```
Comment (davorpml) [inline src/CardRequestService.php:28]: use camelCase please

File: src/CardRequestService.php [hunk 7]
```diff
@@ -5,50 +5,46 @@
...
-                // dd($api_key);
                 $headers = [
                     'AUTH-KEY' => $api_key
                 ];
```
Comment (davorpml) [inline src/CardRequestService.php:31]: check intendation

File: src/CardRequestService.php [hunk 8]
```diff
@@ -5,50 +5,46 @@
...
-                //$params = [];
-        try{      $response = $client->request($method, $url, [
-                    //'json' => $params,
+             $response = $client->request($method, $url, [
```
Comment (davorpml) [inline src/CardRequestService.php:32]: add a blank line above

File: src/CardRequestService.php [hunk 9]
```diff
@@ -5,50 +5,46 @@
...
                     'verify'=>false,
                 ]);
-                 $responseBody = json_decode($response->getBody());
-                return compact('responseBody');
-            }catch(ClientException $ex){
-           // return $responseBody = $ex->getResponse()->getBody(true);
-               // return "TRY block did not response ";
-            }
+            $responseBody = json_decode($response->getBody());
+            return compact('responseBody');
```
Comment (davorpml) [inline src/CardRequestService.php:38]: add a blank line above

File: src/CardRequestService.php [hunk 10]
```diff
@@ -5,50 +5,46 @@
...
+            }catch(RequestException $ex){
```
Comment (davorpml) [inline src/CardRequestService.php:39]: code style:` } catch (RequestException $ex) {`

File: src/CardRequestService.php [hunk 11]
```diff
@@ -5,50 +5,46 @@
...
+              if($ex->hasResponse()){
+                  if($ex->getResponse()->getStatusCode()){
+                      echo "Got response code" . $ex->getResponse()->getStatusCode();
```
Comment (davorpml) [inline src/CardRequestService.php:42]: Missing space after word "code"

File: src/CardRequestService.php [hunk 12]
```diff
@@ -5,50 +5,46 @@
...
-                 $responseBody = json_decode($response->getBody());
-                return compact('responseBody');
-            }catch(ClientException $ex){
-           // return $responseBody = $ex->getResponse()->getBody(true);
-               // return "TRY block did not response ";
-            }
+            $responseBody = json_decode($response->getBody());
+            return compact('responseBody');
+            }catch(RequestException $ex){
+              if($ex->hasResponse()){
```
Comment (davorpml) [inline src/CardRequestService.php:40]: code style: if ($ex->hasResponse()) {

File: src/CardRequestService.php [hunk 13]
```diff
@@ -5,50 +5,46 @@
...
+                  if($ex->getResponse()->getStatusCode()){
```
Comment (davorpml) [inline src/CardRequestService.php:41]: you should put this condition in the first `if`:
`if ($ex->hasResponse() && $ex->getResponse()->getStatusCode()) {`

File: src/CardRequestService.php [hunk 14]
```diff
@@ -5,50 +5,46 @@
...
+                      echo "Got response code" . $ex->getResponse()->getStatusCode();
+                  }
+              }else{
```
Comment (davorpml) [inline src/CardRequestService.php:44]: code style:  `} else {`

File: src/CardRequestService.php [hunk 15]
```diff
@@ -5,50 +5,46 @@
...
+                  throw new \Exception("Request malformd." .$ex->getMessage());
```
Comment (davorpml) [inline src/CardRequestService.php:45]: Missing space after "malformd."

File: src/CardRequestService.php [hunk 16]
```diff
@@ -5,50 +5,46 @@
...
+              }
         }
```
Comment (davorpml) [inline src/CardRequestService.php:48]: remove blank line

File: src/CountryRequestService.php [hunk 17]
```diff
@@ -5,54 +5,64 @@
...
-class CountryRequestService implements IRequestService
+class CountryRequestService extends validateRequest implements IRequestService
 {
-protected $client;
-public function makeRequest($method, $endpointName, $param)
-{
-    //check endpoint from confing by endpointName and create endpointUrl
+    private $modelName = 'country';
+    private $client;
+    protected string $api_key;
```
Comment (davorpml) [inline src/CountryRequestService.php:15]: use camelCase please

File: src/CountryRequestService.php [hunk 18]
```diff
@@ -5,54 +5,64 @@
...
+    protected string $method;
+    protected string $url;
+    protected mixed $param;
 
-    if($endpointName == ''){
-        $url='https://restcountries.com/v2/all';
-    }else{
-        $endUrl = config('endpoints.'.$endpointName);
-        //  dd($endUrl);
+//check endpoint from confing by endpointName and create endpointUrl
```
Comment (davorpml) [inline src/CountryRequestService.php:20]: check intendation and point of the comment

File: src/CountryRequestService.php [hunk 19]
```diff
@@ -5,54 +5,64 @@
...
+     * @param $modelName
+     * @param $method
+     * @param $endpointName
+     * @param null $param
+     * @return string
+     */
+    public function validateRequest($modelName, $method, $endpointName, $param=null)
+    {
+        $requestCheck = new validateRequest();
+        return $this->url = $requestCheck->validate($modelName, $method, $endpointName, $param);
```
Comment (davorpml) [inline src/CountryRequestService.php:32]: ad blank line above return

File: src/CountryRequestService.php [hunk 20]
```diff
@@ -5,54 +5,64 @@
...
+    /**
+     * @param $method
+     * @param $url
+     * @param null $param
+     * @return array|string
+     * @throws \GuzzleHttp\Exception\GuzzleException
+     */
+    public function makeRequest($method, $url, $param=null)
+    {
+         try{
```
Comment (davorpml) [inline src/CountryRequestService.php:44]: code style: `try {`

File: src/CountryRequestService.php [hunk 21]
```diff
@@ -5,54 +5,64 @@
...
+            $client = new Client();
+            $api_key = config('auth_app.AUTH-KEY');
```
Comment (davorpml) [inline src/CountryRequestService.php:46]: use camelCase

File: src/CountryRequestService.php [hunk 22]
```diff
@@ -5,54 +5,64 @@
...
+            $headers = [
+                'AUTH-KEY' => $api_key
+            ];
+            $response = $client->request($method, $url, [
+                'json' => $param,
+                'headers' => $headers,
+                'verify'=>false,
+            ]);
+            $responseBody = json_decode($response->getBody());
+            return compact('responseBody');
```
Comment (davorpml) [inline src/CountryRequestService.php:56]: add blank line above

File: src/CountryRequestService.php [hunk 23]
```diff
@@ -5,54 +5,64 @@
...
+            }catch(RequestException $ex){
```
Comment (davorpml) [inline src/CountryRequestService.php:57]: code style: `} catch(RequestException $ex) {`

File: src/CountryRequestService.php [hunk 24]
```diff
@@ -5,54 +5,64 @@
...
+                if($ex->hasResponse()){
```
Comment (davorpml) [inline src/CountryRequestService.php:58]: code style: `if ($ex->hasResponse()) {`

File: src/CountryRequestService.php [hunk 25]
```diff
@@ -5,54 +5,64 @@
...
+                    if($ex->getResponse()->getStatusCode()){
```
Comment (davorpml) [inline src/CountryRequestService.php:59]: put this condition in first `if`:
`if ($ex->hasResponse() && $ex->getResponse()->getStatusCode()) {`

File: src/CountryRequestService.php [hunk 26]
```diff
@@ -5,54 +5,64 @@
...
+                        echo "Got response code" . $ex->getResponse()->getStatusCode();
```
Comment (davorpml) [inline src/CountryRequestService.php:60]: missing space after word "code" at the end of the string

File: src/CountryRequestService.php [hunk 27]
```diff
@@ -5,54 +5,64 @@
...
+                    }
+                }else{
```
Comment (davorpml) [inline src/CountryRequestService.php:62]: code style: `} else {`

File: src/CountryRequestService.php [hunk 28]
```diff
@@ -5,54 +5,64 @@
...
+                    throw new \Exception("Request malformd." .$ex->getMessage());
```
Comment (davorpml) [inline src/CountryRequestService.php:63]: missing space after word "malformd." at the end of the string

File: src/RequestManager.php [hunk 29]
```diff
@@ -24,7 +24,8 @@ public function make($name):IRequestService{
         }
         $createMethod = 'create'.ucfirst($name).'RequestService';
```
Comment (davorpml) [inline src/RequestManager.php:25]: code style: `$createMethod = 'create' . ucfirst($name) . 'RequestService';`

File: src/config/endpoints.php [hunk 30]
```diff
@@ -1,14 +1,27 @@
...
-    'get the card info' => '/cards/:id',
-    'gets the balance of the card' => '/cards/:id/balance',
-    'gets the pin code of the card' => '/cards/:id/pin',
-    'get a list of transaction history, expects a timeframe' => '/cards/:id/history',
-    'creates a new debit card' => '/cards/create',
-    'deactivates a card'=> '/cards/:id/deactivate',
-    'activates a card' => '/cards/:id/activate',
-    'updates the pin of the card' => '/cards/:id/update',
-    'loads card with a balance, expects to get the amount to load as a parameter' => '/cards/:id/load'
+    'country'=>[
```
Comment (davorpml) [inline src/config/endpoints.php:3]: code style: `'country' => [`

File: src/config/endpoints.php [hunk 31]
```diff
@@ -1,14 +1,27 @@
...
+        'get'=>[
```
Comment (davorpml) [inline src/config/endpoints.php:4]: code style: `'get' => [`

File: src/config/endpoints.php [hunk 32]
```diff
@@ -1,14 +1,27 @@
...
+            'a list of the available countries' => '/countries',
+            'get a specific country' => '/countries/:id',
+        ],
+        'post'=>[]
```
Comment (davorpml) [inline src/config/endpoints.php:8]: code style: `'post' => []`

File: src/config/endpoints.php [hunk 33]
```diff
@@ -1,14 +1,27 @@
...
+    ],
+    'card' =>[
```
Comment (davorpml) [inline src/config/endpoints.php:10]: code style: `'card' => [`

File: src/config/endpoints.php [hunk 34]
```diff
@@ -1,14 +1,27 @@
...
+        'get'=> [
```
Comment (davorpml) [inline src/config/endpoints.php:11]: code style: `'get' => [`

File: src/config/endpoints.php [hunk 35]
```diff
@@ -1,14 +1,27 @@
...
+            'country list' => 'https://countriesnow.space/api/v0.1/countries',
+             'get the card info' => '/cards/:id',
```
Comment (davorpml) [inline src/config/endpoints.php:13]: intendation

File: src/config/endpoints.php [hunk 36]
```diff
@@ -1,14 +1,27 @@
...
+        'post' => [
+            'creates a new debit card' => '/cards/create',
+            'deactivates a card'=> '/cards/:id/deactivate',
+            'activates a card' => '/cards/:id/activate',
+            'updates the pin of the card' => '/cards/:id/update',
+            'loads card with a balance, expects to get the amount to load as a parameter' => '/cards/:id/load'
+        ]
+    ],
+
+
```
Comment (davorpml) [inline src/config/endpoints.php:26]: remove blank lines

File: src/validateRequest.php [hunk 37]
```diff
@@ -0,0 +1,76 @@
...
+    /**
+     * @param $modelName
+     * @param $method
+     * @param $endpointName
+     * @return \Illuminate\Config\Repository|\Illuminate\Contracts\Foundation\Application|mixed
+     */
+    public function readEnd($modelName, $method, $endpointName)
+    {
+       return $point = config('endpoints.'.$modelName.'.'.$method .'.'.$endpointName.'');
+
```
Comment (davorpml) [inline src/validateRequest.php:23]: remove blank line

File: src/validateRequest.php [hunk 38]
```diff
@@ -0,0 +1,76 @@
...
+    /**
+     * @param $modelName
+     * @param $method
+     * @param $endpointName
+     * @return \Illuminate\Config\Repository|\Illuminate\Contracts\Foundation\Application|mixed
+     */
+    public function readEnd($modelName, $method, $endpointName)
+    {
+       return $point = config('endpoints.'.$modelName.'.'.$method .'.'.$endpointName.'');
```
Comment (davorpml) [inline src/validateRequest.php:22]: better like this: `return $point = config("endpoints.{$modelName}.{$method}.{$endpointName}");`

File: src/validateRequest.php [hunk 39]
```diff
@@ -0,0 +1,76 @@
...
+
+    }
+    /**
+     * @param $method
+     * @param $endpointName
+     * @param $param
+     * @return string
+     */
+    public function validate($modelName, $method, $endpointName, $param=null){
+        if($this->getMethod($modelName, $method, $endpointName) == true ){
```
Comment (davorpml) [inline src/validateRequest.php:32]: code style and do you need `== true` ?  `if ($this->getMethod($modelName, $method, $endpointName)) {`

File: src/validateRequest.php [hunk 40]
```diff
@@ -0,0 +1,76 @@
...
+       return $point = config('endpoints.'.$modelName.'.'.$method .'.'.$endpointName.'');
+
+    }
+    /**
+     * @param $method
+     * @param $endpointName
+     * @param $param
+     * @return string
+     */
+    public function validate($modelName, $method, $endpointName, $param=null){
```
Comment (davorpml) [inline src/validateRequest.php:31]: code style, add space between: `$param=null) {`

File: src/validateRequest.php [hunk 41]
```diff
@@ -0,0 +1,76 @@
...
+        if($this->getMethod($modelName, $method, $endpointName) == true ){
+           return $this->getUrl($this->getMethod($modelName, $method, $endpointName));
+        }else{
```
Comment (davorpml) [inline src/validateRequest.php:34]: code style: `} else {`

File: src/validateRequest.php [hunk 42]
```diff
@@ -0,0 +1,76 @@
...
+
+    /**
+     * @param $point
+     * @param null $param
+     * @return string
+     * @description check endpoint regularity
+     */
+    public function getUrl($point, $param=null)
+    {
+            if(strpos($point, ':id') !==false && (!empty($param))){
```
Comment (davorpml) [inline src/validateRequest.php:47]: code style: `if (strpos($point, ':id') !== false && (!empty($param))) {`

File: src/validateRequest.php [hunk 43]
```diff
@@ -0,0 +1,76 @@
...
+                $endUrl = str_replace(':id', $param,$point);
```
Comment (davorpml) [inline src/validateRequest.php:48]: add space after comma

File: src/validateRequest.php [hunk 44]
```diff
@@ -0,0 +1,76 @@
...
+            }elseif(strpos($point, ':id') !==false && (empty($param))){
```
Comment (davorpml) [inline src/validateRequest.php:49]: code style:` } elseif(strpos($point, ':id') !== false && (empty($param))) {`

File: src/validateRequest.php [hunk 45]
```diff
@@ -0,0 +1,76 @@
...
+                return "Missing param for ".$point;
```
Comment (davorpml) [inline src/validateRequest.php:50]: code style ` return "Missing param for " . $point;`

File: src/validateRequest.php [hunk 46]
```diff
@@ -0,0 +1,76 @@
...
+            }else{
```
Comment (davorpml) [inline src/validateRequest.php:51]: code style `} else {`

File: src/validateRequest.php [hunk 47]
```diff
@@ -0,0 +1,76 @@
...
+                $endUrl = $point;
+            }
+            $baseURL = env('APP_URL');
+            $this->url = $baseURL.$endUrl;
```
Comment (davorpml) [inline src/validateRequest.php:55]: check indentation for this line and all lines above
code style: `$this->url = $baseURL . $endUrl;`

File: src/validateRequest.php [hunk 48]
```diff
@@ -0,0 +1,76 @@
...
+    }
+
+    /**
+     * @param $method
+     * @param $endpointName
+     * @return \Illuminate\Config\Repository|\Illuminate\Contracts\Foundation\Application|mixed|string
+     */
+    public function getMethod($modelName, $method, $endpointName)
+    {
+        $point = config('endpoints.'.$modelName.'.'.$method .'.'.$endpointName.'');
```
Comment (davorpml) [inline src/validateRequest.php:67]: `$point = config("endpoints.{$modelName}.{$method}.{$endpointName}");`

File: src/validateRequest.php [hunk 49]
```diff
@@ -0,0 +1,76 @@
...
+
+       if(!$point){
```
Comment (davorpml) [inline src/validateRequest.php:69]: code style: `if (!$point) {`

File: src/validateRequest.php [hunk 50]
```diff
@@ -0,0 +1,76 @@
...
+          return  "Bad method or modelName for endpoint Name.";
+       }else{
```
Comment (davorpml) [inline src/validateRequest.php:71]: code style: `} else {`

File: src/ReadMe [hunk 51]
```diff
@@ -27,27 +15,21 @@ koriscenje high level klase npr. u controller-u
...
+@params
+    $modelName= 'country';
+    $method = 'get';
+    $endpointName = 'a list of the available countries';
+    $param = '1';
 
  $factory = app(IRequestManager::class);
         $service = $factory->make($modelName);
-        $response = $service->makeRequest($method, $endpointName, $param);
+        $url = $srvice->validate($modelName, mathod, $endpointName);
```
Comment (davorpml) [inline src/ReadMe:26]: Please be precise `$srvice` ?              `mathod` ??

**Review by davorpml — COMMENTED** [review]
Hello Marijana,

You were interviewed by my colleague and dev team leader Simo Jaković.
Because you are a potential reinforcement in my team, I added a few comments. Check it, please.

Have a great rest of the week and holidays!

All the best,

Davor

**Comment by marijanafilipovic** [comment]
> Hello Marijana,
> 
> You were interviewed by my colleague and dev team leader Simo Jaković. Because you are a potential reinforcement in my team, I added a few comments. Check it, please.
> 
> Have a great rest of the week and holidays!
> 
> All the best,
> 
> Davor

**Comment by marijanafilipovic** [comment]
Hello Davor, 

Thank you for your comments. I agree with it, all. I will do the changes.
Have a great holidey time! 

All the best,
Marijana

**Pull request merged**